### DISCUSSION TITLE
Shape3D doesn't have a collide method

### DISCUSSION THREAD

**Discussion by Calandiel** [godotengine/godot-proposals#8406] [3D]
Compare Shape2D API with that of Shape3D.
The latter has fewer methods. One particularly problematic omission is the collide method (that checks for a collision between two shapes given their transforms). Currently, it doesn't seem to be possible to do without the use of rather odd node based setups (or doing intersect shape while filtering out all RIDs besides the shapes in question), thus incurring costs of allocating those nodes and making optimization with PhysicsServer3D more difficult than it ought to be.

Is there any reason for why it's missing? At a glance, the implementation from Shape2D should be relatively easy to port to Shape3D, no?

**Comment by AThousandShips** [comment]
As far as I know the 2D one is not used generally in the engine, and the 3D case isn't needed, it appears the 2D case is only used for `TouchScreenButton` which would explain the existence of the method

**Reply by AThousandShips** [reply]
> At a glance, the implementation from Shape2D should be relatively easy to port to Shape3D, no?

Not really, the required method `shape_collide` doesn't exist in the 3D physics server, so it'd have to be implemented, and unsure if it is trivial

**Reply by Calandiel** [reply]
Could you share why it's not needed?
Surely, methods that aren't used by the engine internally are still useful, no?

**Reply by AThousandShips** [reply]
It's not needed for the engine, is what I mean, so it hasn't been added, the 2D case seems to have been added for the engine to use and not added specifically for script use

**Reply by Calandiel** [reply]
Right 🤔 
Thats a p big performance bottleneck for anything thats stressing the physics server, though.
For example, implementing custom navmesh generation without having a way to check for collision between shapes adds the overhead of dealing with bodies.

Well, I suppose it won't be the first time I replace a large part of the engine with something external.

**Reply by AThousandShips** [reply]
I'd suggest writing a proper proposal and present your arguments and cases for it and suggestions for it to be added

**Reply by Calandiel** [reply]
Sadly, I need it sooner than the formal process would take and my time is very limited these days so I'll have to settle for making a custom solution for now.

**Reply by AThousandShips** [reply]
Okay, if you don't feel interested in promoting the idea for everyone then you'll have to settle for a workaround I guess

But there's no other way to add this feature than making a proposal, so I'm confused why you posted this in the first place if you can't wait?

**Reply by Calandiel** [reply]
Custom solutions are hardly workarounds and pushing for an implementation of something into an open source project without being previously associated with it often takes longer than doing the thing yourself.

As to why I asked the question, I wanted my original question answered. That is:
```
Is there any reason for why it's missing? At a glance, the implementation from Shape2D should be relatively easy to port to Shape3D, no?
```

**Reply by AThousandShips** [reply]
And I answered it, and suggested what I suggested, this is not for general support questions but for proposals for the engine, hence my confusion, I assumed you were aware of that and that your intention was to discuss improvements to the engine :slig

**Reply by Calandiel** [reply]
It was, but I was working with the assumption that adding it would be a formality.
To quote you:
```
Not really, the required method shape_collide doesn't exist in the 3D physics server, so it'd have to be implemented, and unsure if it is trivial
```
I did take a second look since your comment and it does look trickier than I first thought.

**Reply by AThousandShips** [reply]
Well I understand, but the original aspect about intent still stands

What's the problem of writing a proposal? It's not a massive amount of work and you'd help everyone if it is supported and popular, and I'm just assuming it's not rivial to add the required step, my comment was simply that it wasn't just a matter of exposing a method

If you do find the time you don't need long to just write a proposal, but it's up to you 🙂 

In either case this can be closed as the question is answered and the option would be either to create a formal proposal, or leave it be

**Reply by Calandiel** [reply]
My understanding is that a proposal ought to suggest an implementation (as that's what the template for issues is called). Is that not the case?

I can write a summary of my search before deciding to "do it myself" but researching a feasible way to implement it within existing 3d physics in Godot is potentially a lot more work than writing a proposal itself.

**Reply by AThousandShips** [reply]
Referencing the 2D case would be enough as an implementation suggestion, it doesn't have to be a complete idea, but you'd need a basic idea, in this case I'd say "make something equivalent of the 2D case" is sufficient

A proposal is more suited to describe the use case and need, and to investigate it, leaving room to discuss and investigate 🙂
